### ISSUE TITLE
[ApiPromise] send() subscription is returning undefined values in addition valid ones

### ISSUE THREAD

**Issue by Stefie** [polkadot-js/api#598]
It looks like trackStatus inside send() is being called twice.

**Comment by Stefie** [comment]
I found the error, but unfortunately I'm not able to resolve it. It seems like some kind of chaining/ event stack issue or unwanted side effect related to  resolve() inside the tap() callback.

The Bug is here:
https://github.com/polkadot-js/api/blob/master/packages/api/src/promise/index.ts#L227
```
tap((params) => {
            if (!isCompleted) {
              isCompleted = true;

              resolve(() => subscription.unsubscribe());
            }
          })
```

If I remove the `resolve(() => subscription.unsubscribe());` there are no additional  values emitted on send(), when I leave it there it does. It has nothing to do with the unsubscribe function, what triggers it is the resolve() .

When I'm logging the value that is piped into tap() I'm first getting the expected value and then the unsubscribe function itself. Twice when I'm using signAndSend() and only once when I'm using sign().send(). That corresponds with the number of calls to the onCall method of  the ApiPromise class... but that might as well be totally unrelated.


![image](https://user-images.githubusercontent.com/125398/51333219-91270b00-1a7d-11e9-883e-16a1ec39acd0.png)

@amaurymartiny I could use a second pair of eyes here :)

**Comment by amaurymartiny** [comment]
Hmm yeah, that code is quite tangled.

How about something in those lines?
```javascript
    return new Promise((resolve, reject) => {
      let isCompleted = false;
      const subscription = method(...remainingParams)
        .subscribe(
          (value) => {
            if (!isCompleted) { isCompleted = true; resolve(()=> subscription); }
            cb(value);
          },
          (err) => reject(err)
       );
}) as PromiseSubscription;
```

Also, @jacogr, why do you actually need to return the unsubscribe function only after 1st value? The simplest would be `const sub = method(...).subscribe(cb); return Promise.resolve(sub.unsubscribe);`.

**Comment by jacogr** [comment]
Reject is the issue. We need to reject on initial error, before result. (In your example above, we may actually resolve and then reject.)

Happy without a tap, reject just needs the isCompleted as well

**Comment by amaurymartiny** [comment]
^ah right, so my proposal won't work.

The "logged twice" problem might come from 
https://github.com/polkadot-js/api/blob/bf40c427f26e72f230161cfc8f4597287007c8b8/packages/api/src/SubmittableExtrinsic.ts#L76

I actually don't think there should be a `statusCb` here. Though not 100% sure, need to test.

**Comment by jacogr** [comment]
Line 79 seems like the issue now that you mention it... https://github.com/polkadot-js/api/blob/bf40c427f26e72f230161cfc8f4597287007c8b8/packages/api/src/SubmittableExtrinsic.ts#L79

**Comment by amaurymartiny** [comment]
Actually the code above will work:
- if 1st value is good, then resolve
- if 1st value is error, then reject

**Comment by Stefie** [comment]
@amaurymartiny 
Unfortunately Your suggestion in  doesn't solve it :|

> Hmm yeah, that code is quite tangled.
> 
> How about something in those lines?
> 
> ```js
>     return new Promise((resolve, reject) => {
>       let isCompleted = false;
>       const subscription = method(...remainingParams)
>         .subscribe(
>           (value) => {
>             if (!isCompleted) { isCompleted = true; resolve(()=> subscription); }
>             cb(value);
>           },
>           (err) => reject(err)
>        );
> }) as PromiseSubscription;
> ```
> 
> Also, @jacogr, why do you actually need to return the unsubscribe function only after 1st value? The simplest would be `const sub = method(...).subscribe(cb); return Promise.resolve(sub.unsubscribe);`.

Unfortunately this doesn't solve it, the bug remains.

**Comment by jacogr** [comment]
It needs to reject on error.

So basically, if you pass in invalid values (e.g. no value to `system.accountNonce`, creation throws - it needs to reject at that point.

**Comment by Stefie** [comment]
And since it also happens on sign() and not only on signAndSend(), I don't think that it has something to do with the callback in api/packages/api/src/SubmittableExtrinsic.ts, or at least not exclusively...

**Comment by jacogr** [comment]
I think it is line 79 - give it a dummy callback and check.

i.e. `[statusCb]` changes to `[() => void 0]`

**Comment by Stefie** [comment]
And, taken from my first comment:

> [It happens] Twice when I'm using signAndSend() and only once when I'm using sign().send(). That corresponds with the number of calls to the onCall method of the ApiPromise class

That sounds a little too related to just be a random coincident...

**Comment by jacogr** [comment]
... your comment points to line 79 :)

**Comment by Stefie** [comment]
Might be... it would be 79 and 59, then. Hang on...

**Comment by Stefie** [comment]
hmmm... nope. I'm not able to resolve it there...

**Comment by jacogr** [comment]
Add logs at both places where it gets passed. See what fires. (Same with the resolve/reject), understand the flow first.

**Comment by jacogr** [comment]
Ok, so I believe both the `[statusCb]` in `send` and `signAndSend` should not be there. Using the e2e promise tx tests with the above logging enabled, getting this - (look specifically at the value passed through on the sign and signAndSend callbacks - anyway, it is a status callback, it should only yield values as part of the tracking)

# sign and send (seperately)

```
console.log packages/api/test/e2e/promise-tx.spec.js:31
    Transaction status: Ready

  console.error packages/api/src/SubmittableExtrinsic.ts:59
    send () => {
                /* istanbul ignore next */
                cov_seoc7xqvo.f[11]++;
                cov_seoc7xqvo.s[25]++;
                return subscription.unsubscribe();
              }

  console.log packages/api/test/e2e/promise-tx.spec.js:31
    Transaction status: undefined

  console.error packages/api/src/SubmittableExtrinsic.ts:50
    trackStatus { events:
       [ EventRecord [Map] {
           'phase' => [Phase],
           'event' => [Event],
           _jsonMap: Map {},
           _Types: [Object] },
         EventRecord [Map] {
           'phase' => [Phase],
           'event' => [Event],
           _jsonMap: Map {},
           _Types: [Object] } ],
      status:
       ExtrinsicStatus {
         raw:
          Finalised [Uint8Array] [
            170,
            142,
            206,
            208,
            17,
            158,
            7,
            46,
            51,
            27,
            183,
            145,
            86,
            211,
            11,
            37,
            12,
            254,
            95,
            243,
            157,
            8,
            35,
            192,
            168,
            224,
            235,
            55,
            31,
            141,
            25,
            243 ],
         _def:
          { Future: [Function: Future],
            Ready: [Function: Ready],
            Finalised: [Function: Finalised],
            Usurped: [Function: Usurped],
            Broadcast: [Function: Broadcast],
            Dropped: [Function: Dropped],
            Invalid: [Function: Invalid] },
         _index: 2,
         _indexes: [ 0, 1, 2, 3, 4, 5, 6 ] },
      type: 'Finalised' }

  console.log packages/api/test/e2e/promise-tx.spec.js:31
    Transaction status: Finalised

  console.log packages/api/test/e2e/promise-tx.spec.js:34
    Completed at block hash 0xaa8eced0119e072e331bb79156d30b250cfe5ff39d0823c0a8e0eb371f8d19f3

  console.log packages/api/test/e2e/promise-tx.spec.js:35
    Events:

  console.log packages/api/test/e2e/promise-tx.spec.js:38
    	 ApplyExtrinsic : balances.Transfer ["5GQATTPFqkfze7kbGuvezhV921FwSp6Xyr8FMW1pmi6LjDuu","5DkQbYAExs3M2sZgT1Ec3mKfZnAQCL4Dt9beTCknkCUn5nni","0x00000000000000000000000000003039","0x00000000000000000000000000000000"]

  console.log packages/api/test/e2e/promise-tx.spec.js:38
    	 ApplyExtrinsic : system.ExtrinsicSuccess []

/// wtf?
  console.error packages/api/src/SubmittableExtrinsic.ts:59
    send () => {
                /* istanbul ignore next */
                cov_seoc7xqvo.f[11]++;
                cov_seoc7xqvo.s[25]++;
                return subscription.unsubscribe();
              }

  console.log packages/api/test/e2e/promise-tx.spec.js:31
    Transaction status: undefined
```

# sign nd send (combined)

```
console.error packages/api/src/SubmittableExtrinsic.ts:59
    send () => {
                /* istanbul ignore next */
                cov_seoc7xqvo.f[11]++;
                cov_seoc7xqvo.s[25]++;
                return subscription.unsubscribe();
              }

  console.log packages/api/test/e2e/promise-tx.spec.js:52
    Transaction status: undefined

  console.error packages/api/src/SubmittableExtrinsic.ts:79
    signAndSend () => {
                /* istanbul ignore next */
                cov_seoc7xqvo.f[11]++;
                cov_seoc7xqvo.s[25]++;
                return subscription.unsubscribe();
              }

  console.log packages/api/test/e2e/promise-tx.spec.js:52
    Transaction status: undefined

  console.error packages/api/src/SubmittableExtrinsic.ts:50
    trackStatus { events:
       [ EventRecord [Map] {
           'phase' => [Phase],
           'event' => [Event],
           _jsonMap: Map {},
           _Types: [Object] },
         EventRecord [Map] {
           'phase' => [Phase],
           'event' => [Event],
           _jsonMap: Map {},
           _Types: [Object] } ],
      status:
       ExtrinsicStatus {
         raw:
          Finalised [Uint8Array] [
            130,
            165,
            102,
            219,
            178,
            14,
            153,
            67,
            249,
            220,
            148,
            194,
            116,
            50,
            132,
            91,
            13,
            7,
            159,
            126,
            236,
            219,
            13,
            109,
            107,
            219,
            153,
            158,
            0,
            6,
            156,
            6 ],
         _def:
          { Future: [Function: Future],
            Ready: [Function: Ready],
            Finalised: [Function: Finalised],
            Usurped: [Function: Usurped],
            Broadcast: [Function: Broadcast],
            Dropped: [Function: Dropped],
            Invalid: [Function: Invalid] },
         _index: 2,
         _indexes: [ 0, 1, 2, 3, 4, 5, 6 ] },
      type: 'Finalised' }

  console.log packages/api/test/e2e/promise-tx.spec.js:52
    Transaction status: Finalised

  console.log packages/api/test/e2e/promise-tx.spec.js:55
    Completed at block hash 0x82a566dbb20e9943f9dc94c27432845b0d079f7eecdb0d6d6bdb999e00069c06

  console.log packages/api/test/e2e/promise-tx.spec.js:56
    Events:

  console.log packages/api/test/e2e/promise-tx.spec.js:59
    	 ApplyExtrinsic : balances.Transfer ["5GQATTPFqkfze7kbGuvezhV921FwSp6Xyr8FMW1pmi6LjDuu","5DkQbYAExs3M2sZgT1Ec3mKfZnAQCL4Dt9beTCknkCUn5nni","0x00000000000000000000000000003039","0x00000000000000000000000000000000"]

  console.log packages/api/test/e2e/promise-tx.spec.js:59
    	 ApplyExtrinsic : system.ExtrinsicSuccess []

  console.error packages/api/src/SubmittableExtrinsic.ts:59
    send () => {
                /* istanbul ignore next */
                cov_seoc7xqvo.f[11]++;
                cov_seoc7xqvo.s[25]++;
                return subscription.unsubscribe();
              }

  console.log packages/api/test/e2e/promise-tx.spec.js:52
    Transaction status: undefined
```

**Comment by Stefie** [comment]
Thanks for investigating! But I found the Bug an hour ago and postponed the Fix to tomorrow.  I didn't know you were still looking, otherwise I would have told you.

The callback is fine. There's the optional bool parameter `isSubscription` on the onCall method of ApiPromise. It has no default value and we don't pass it to any of the function calls to onCall in `SubmittableExtrinsic.ts` (and possibly other parts of the code, that's why I postponed fixing it to tomorrow). That way it's undefined  for `send()`, `sign()`, `sendAndSign()` and `trackStatus()` instead of true or false and the onCall function can't do it's job properly.

https://github.com/polkadot-js/api/blob/master/packages/api/src/promise/index.ts#L196

The fix here is to pass `isSubscription === false` for  `sign()`, `sendAndSign()` and `trackStatus()` and only true for `send()` or just set a default value.

I'll fix it first thing tomorrow morning!

**Comment by Stefie** [comment]
So for example this in `src/promise/index.js`
`protected onCall (method: RxFn, params: Array<any>, isSubscription?: boolean = false): OnCall {}`

and then this in `SubmittableExtrinsic.ts`:
```
  send (statusCb?: (result: SubmittableSendResult) => any): Observable<SubmittableSendResult> {
    return this._onCall(
      () => (this._api.rpc.author
        .submitAndWatchExtrinsic(this) as Observable<ExtrinsicStatus>)
        .pipe(switchMap((status) => this.trackStatus(status, statusCb))),
      [statusCb], 
      true
    ) as unknown as Observable<SubmittableSendResult>;
  }
```

**Comment by jacogr** [comment]
I actually ran into exactly this on apps when I upgraded to use derive - getting an invalid/empty events structure back. (Had bigger problems to deal with, so didn't look into it at all.)

Just be careful with the send - when no statusCb is provided to it, it should return a promise, not a subscription promise. (Not 100% sure we are currently handling that case correctly - we don't have e2e tests for it at all - as per the original impl. at that point it uses submitExtrinsic - I don't think we have this feature atm, which is an oversight)

**Comment by jacogr** [comment]
Yeap, seems to affect apps in weird ways - just got something from @chevdor where status type is not sent along and apps assumes it has it and tries to work with it. So I didn't manage o catch them all with workarounds - so things are breaking atm.

**Comment by Stefie** [comment]
> Just be careful with the send - when no statusCb is provided to it, it should return a promise, not a subscription promise.

That's already being handled in the ApiPromise `onCall()`, but I'll double-check!

One of my guesses is that because we're using switchMap(), all the observables with ìsSubscription === undefined` are initialized and subscribed to and then immediately canceled and replaced by the next subscription before receiving a value. And the tap() function just caught/ logged this behaviour

**Comment by jacogr** [comment]
Going to close this one since we have a quickfix in place via #601, will replace and link with an bubble-up-isSubscription version
